client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java

```diff
@@ -1588,9 +1519,169 @@ private void closeChannel() {
                 LoggerHelpers.traceLeave(log, "getCurrentSegmentsForKeyValueTable", traceId);
             });
         }
-
     //endregion
 
+    // region ReaderGroups
+    public CompletableFuture<Boolean> createReaderGroup(String scope, String rgName, final ReaderGroupConfig rgConfig) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        Exceptions.checkNotNullOrEmpty(scope, "scope");
+        Exceptions.checkNotNullOrEmpty(rgName, "rgName");
+        Preconditions.checkNotNull(rgConfig, "rgConfig");
+        final long requestId = requestIdGenerator.get();
+        long traceId = LoggerHelpers.traceEnter(log, "createReaderGroup", rgConfig, requestId);
+
+        final CompletableFuture<CreateReaderGroupStatus> result = this.retryConfig.runAsync(() -> {
+            RPCAsyncCallback<CreateReaderGroupStatus> callback = new RPCAsyncCallback<>(requestId, "createReaderGroup", scope, rgName, rgConfig);
+            new ControllerClientTagger(client, timeoutMillis).withTag(requestId, "createReaderGroup", scope, rgName)
+                    .createReaderGroup(ModelHelper.decode(scope, rgName, rgConfig), callback);
+            return callback.getFuture();
+        }, this.executor);
+        return result.thenApply(x -> {
+            switch (x.getStatus()) {
+                case FAILURE:
+                    log.warn(requestId, "Failed to create reader group: {}", rgName);
+                    throw new ControllerFailureException("Failed to create reader group: " + rgName);
+                case INVALID_RG_NAME:
+                    log.warn(requestId, "Illegal Reader Group Name: {}", rgName);
+                    throw new IllegalArgumentException("Illegal readergroup name: " + rgName);
+                case SCOPE_NOT_FOUND:
+                    log.warn(requestId, "Scope not found: {}", scope);
+                    throw new IllegalArgumentException("Scope does not exist: " + scope);
+                case SUCCESS:
+                    log.info(requestId, "ReaderGroup created successfully: {}", rgName);
+                    return true;
+                case UNRECOGNIZED:
+                default:
+                    throw new ControllerFailureException("Unknown return status creating reader group " + rgName
+                            + " " + x.getStatus());
+            }
+        }).whenComplete((x, e) -> {
+            if (e != null) {
+                log.warn(requestId, "createReaderGroup {}/{} failed: ", scope, rgName, e);
+            }
+            LoggerHelpers.traceLeave(log, "createReaderGroup", traceId, rgConfig, requestId);
+        });
+    }
+
+    public CompletableFuture<Boolean> updateReaderGroup(String scope, String rgName, final ReaderGroupConfig rgConfig) {
```

Comment:
please open an issue to return generation back from controller.

Reply:
Filed: https://github.com/pravega/pravega/issues/5447